hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/audit/mapreduce/S3AAuditLogMergerAndParser.java

```diff
@@ -0,0 +1,237 @@
+    boolean patternMatching = matcher.matches();
+    if (patternMatching) {
+      for (String key : AWS_LOG_REGEXP_GROUPS) {
+        try {
+          final String value = matcher.group(key);
+          auditLogMap.put(key, value);
+        } catch (IllegalStateException e) {
+          LOG.debug(String.valueOf(e));
+        }
+      }
+    }
+    return auditLogMap;
+  }
+
+  /**
+   * parseReferrerHeader method helps in parsing the http referrer header.
+   * which is one of the key-value pair of audit log
+   *
+   * @param referrerHeader this is the http referrer header of a particular audit log
+   * @return it returns a map i.e, auditLogMap which contains key-value pairs
+   * of audit log as well as referrer header present in it
+   */
+  public HashMap<String, String> parseReferrerHeader(String referrerHeader) {
+    HashMap<String, String> referrerHeaderMap = new HashMap<>();
+    if (referrerHeader == null || referrerHeader.length() == 0) {
+      LOG.debug(
+          "This is an empty string or null string, expected a valid string to parse");
+      return referrerHeaderMap;
+    }
+    int indexOfQuestionMark = referrerHeader.indexOf("?");
+    String httpReferrer = referrerHeader.substring(indexOfQuestionMark + 1,
+        referrerHeader.length() - 1);
+    int lengthOfReferrer = httpReferrer.length();
+    int start = 0;
+    while (start < lengthOfReferrer) {
+      int equals = httpReferrer.indexOf("=", start);
+      // no match : break
+      if (equals == -1) {
+        break;
+      }
+      String key = httpReferrer.substring(start, equals);
+      int end = httpReferrer.indexOf("&", equals);
+      // or end of string
+      if (end == -1) {
+        end = lengthOfReferrer;
+      }
+      String value = httpReferrer.substring(equals + 1, end);
+      referrerHeaderMap.put(key, value);
+      start = end + 1;
+    }
+    return referrerHeaderMap;
+  }
+
+  /**
+   * Merge and parse all the audit log files and convert data into avro file.
+   *
+   * @param fileSystem filesystem
+   * @param logsPath   source path of logs
+   * @param destPath   destination path of merged log file
+   * @return true
+   * @throws IOException on any failure
+   */
+  public boolean mergeAndParseAuditLogFiles(FileSystem fileSystem,
+      Path logsPath,
+      Path destPath) throws IOException {
+
+    // Listing file in given path
+    RemoteIterator<LocatedFileStatus> listOfLogFiles =
+        fileSystem.listFiles(logsPath, true);
+
+    Path destFile = new Path(destPath, "AuditLogFile");
+
+    try (FSDataOutputStream fsDataOutputStream = fileSystem.create(destFile)) {
+
+      // Iterating over the list of files to merge and parse
+      while (listOfLogFiles.hasNext()) {
+        FileStatus fileStatus = listOfLogFiles.next();
+        int fileLength = (int) fileStatus.getLen();
+        byte[] byteBuffer = new byte[fileLength];
+        try (FSDataInputStream fsDataInputStream =
+            fileSystem.open(fileStatus.getPath())) {
```

Comment:
can you use the new openFile builder; there are examples such as
org.apache.hadoop.fs.AvroFSInput#AvroFSInput(org.apache.hadoop.fs.FileContext, org.apache.hadoop.fs.Path) to show the basic strategy.

if you an add the filestatus in .withFileStatus() we can save a HEAD request opening files on s3 and abfs, for faster reads

Reply:
modified as suggested